terraform/service_mirror.tf

```diff
@@ -0,0 +1,195 @@
+
+  secret_id = google_secret_manager_secret.db-secret[each.key].id
+  role      = "roles/secretmanager.secretAccessor"
+  member    = "serviceAccount:${google_service_account.mirror.email}"
+}
+
+resource "google_project_iam_member" "mirror-observability" {
+  for_each = toset([
+    "roles/cloudtrace.agent",
+    "roles/logging.logWriter",
+    "roles/monitoring.metricWriter",
+    "roles/stackdriver.resourceMetadata.writer",
+  ])
+
+  project = var.project
+  role    = each.key
+  member  = "serviceAccount:${google_service_account.mirror.email}"
+}
+
+resource "google_cloud_run_service" "mirror" {
+  name     = "mirror"
+  location = var.cloudrun_location
+
+  autogenerate_revision_name = true
+
+  template {
+    spec {
+      service_account_name = google_service_account.mirror.email
+
+      containers {
+        image = "gcr.io/${data.google_project.project.project_id}/github.com/google/exposure-notifications-server/mirror:initial"
+
+        resources {
+          limits = {
+            cpu    = "2000m"
+            memory = "1G"
+          }
+        }
+
+        dynamic "env" {
+          for_each = merge(
+            local.common_cloudrun_env_vars,
+
+            // This MUST come last to allow overrides!
+            lookup(var.service_environment, "mirror", {}),
+          )
+
+          content {
+            name  = env.key
+            value = env.value
+          }
+        }
+      }
+
+      container_concurrency = 5
+      // 30 seconds less than cloud scheduler maximum.
+      timeout_seconds = 570
+    }
+
+    metadata {
+      annotations = {
+        "autoscaling.knative.dev/maxScale" : "10",
+        "run.googleapis.com/vpc-access-connector" : google_vpc_access_connector.connector.id
+      }
+    }
+  }
+
+  depends_on = [
+    google_project_service.services["run.googleapis.com"],
+    google_secret_manager_secret_iam_member.mirror-db,
+    null_resource.build,
+    null_resource.migrate,
+  ]
+
+  lifecycle {
+    ignore_changes = [
+      template[0].metadata[0].annotations,
+      template[0].spec[0].containers[0].image,
+    ]
+  }
+}
+
+
+#
+# Create scheduler job to invoke the service on a fixed interval.
+#
+
+resource "google_service_account" "mirror-invoker" {
+  project      = data.google_project.project.project_id
+  account_id   = "en-mirror-invoker-sa"
+  display_name = "Exposure Notification Mirror Invoker"
+}
+
+resource "google_cloud_run_service_iam_member" "mirror-invoker" {
+  project  = google_cloud_run_service.mirror.project
+  location = google_cloud_run_service.mirror.location
+  service  = google_cloud_run_service.mirror.name
+  role     = "roles/run.invoker"
+  member   = "serviceAccount:${google_service_account.mirror-invoker.email}"
+}
+
+resource "google_cloud_scheduler_job" "mirror-worker" {
+  name             = "mirror-worker"
+  region           = var.cloudscheduler_location
+  schedule         = "* * * * *"
+  time_zone        = "America/Los_Angeles"
+  attempt_deadline = "600s"
```

Comment:
With a deadline of 600s, and a schedule of every minute, this could have up to ten running at the same time, is that ok?

Reply:
yes - that's intentional.

Any given mirror instance holds a lock while it's running.